File: archive/2026-04-05/ch06_tree/t16_path_sum.py
"""LeetCode 113. Path Sum II
https://leetcode.cn/problems/path-sum-ii/
"""

from typing import List, Optional

from tree_node import TreeNode


class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        if not root:
            return []

        path: List[int] = []
        ans: List[List[int]] = []

        def traversal(root: Optional[TreeNode], targetSum: int) -> None:
            if not root.left and not root.right:
                if targetSum == root.val:
                    path.append(root.val)
                    ans.append(path.copy())
                    path.pop()
                return

            path.append(root.val)
            if root.left:
                traversal(root.left, targetSum - root.val)
            if root.right:
                traversal(root.right, targetSum - root.val)
            path.pop()

        traversal(root, targetSum)
        return ans


if __name__ == "__main__":
    solution = Solution()
    root1 = TreeNode(
        5,
        TreeNode(4, TreeNode(11, TreeNode(7), TreeNode(2))),
        TreeNode(8, TreeNode(13), TreeNode(4, TreeNode(5), TreeNode(1))),
    )
    result1 = sorted(solution.pathSum(root1, 22))
    expected = sorted([[5, 4, 11, 2], [5, 8, 4, 5]])
    assert result1 == expected

    root2 = TreeNode(1, TreeNode(2), TreeNode(3))
    assert solution.pathSum(root2, 5) == []

    root3 = TreeNode(1, TreeNode(2))
    assert solution.pathSum(root3, 0) == []
    print("All tests passed.")

File: archive/2025-04-04/ch06_tree/T16_pathSum.java
package org.review2025spring.ch06_tree;

import java.util.ArrayList;
import java.util.List;

public class T16_pathSum {
    List<List<Integer>> res = new ArrayList<>();
    List<Integer> path = new ArrayList<>();

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        if (root == null) return res;
        traversal(root, targetSum);
        return res;
    }

    public void traversal(TreeNode root, int sum) {
        if (root.left == null && root.right == null) {
            if (sum == root.val) {
                path.add(root.val);
                res.add(new ArrayList<>(path));
                path.remove(path.size() - 1);
            }
            return;
        }
        path.add(root.val);
        if (root.left != null) traversal(root.left, sum - root.val);
        if (root.right != null) traversal(root.right, sum - root.val);
        path.remove(path.size() - 1);
    }
}
